guides/source/testing.md

```diff
@@ -1398,6 +1398,56 @@ class ProfileControllerTest < ActionDispatch::IntegrationTest
 end
 ```
 
+#### Extracting Files
+
+If you find your helpers are clutering `test_helper.rb`, you can extract them into separate files. One good place to store them is `lib/test`.
+
+```ruby
+# lib/test/multiple_assertions.rb
+module MultipleAssertions
+  def assert_multiple_of_fourty_two(number)
+    assert (number % 42 == 0), 'expected #{number} to be a multiple of 42'
+  end
+end
+```
+
+These helpers can then be explicitly required as needed and included as needed
+
+```ruby
+require 'test_helper'
+require Rails.root.join('lib', 'test', 'multiple_assertions')
+
+class NumberTest < ActiveSupport::TestCase
+  include MultipleAssertions
+
+  test '420 is a multiple of fourty two' do
+    assert_multiple_of_fourty_two 420
+  end
+end
+```
+
+or they can continue to be included directly into the relevant parent classes
+
+```ruby
+# test/test_helper.rb
+require Rails.root.join('lib', 'test', 'sign_in_helper')
+
+class ActionDispatch::IntegrationTest
```

Comment:
Use lazy load hook instead

```ruby
ActiveSupport.on_load(: action_dispatch_integration_test) do
  include SignInHelper
end
```

Reply:
Interesting. I hadn't come across this before. Is there much benefit here, since by this point `test_helper` has required the Rails environment already.

I can switch to this if it's the preferred pattern, but I basically just copied the code from earlier in the same guide. Should I switch it in both places?

Reply:
I've never seen this used for test code, but I'm not against it. In your test helper, its fairly safe to assume that `ActionDispatch::IntegrationTest` is loaded, I think.

Reply:
https://github.com/rails/rails/blob/5bb4546439ed20a864c6df18788dce0eb45d8326/railties/lib/rails/test_help.rb

Reply:
I could open a follow up PR where I update all the places that re-open `IntegrationTest` to be consistent (in this file, and in the generated test helper).

Reply:
Sorry! Forgot to answer 😅 . 

I think the generated_test_helper.rb isn't up to date, but anyway using a lazy load hook for ActiveSupport::TestCase or even ActionDispatch::IntegrationTest isn't a big deal compare to using a lazy_load_hook for ActiveRecord or ActionController for example, so I'm fine with either :).

@sambostock Feel free to open another PR to make everything consistent

Reply:
Yes, it is a big deal. `ActionDispatch::IntegrationTest` loads `ActionController::Base`. For test_helpers the current syntax is fine since this is after the environment is loaded and the application is initialized, but for gems it is a problem.

Reply:
Addressed in #34619.
- If this is merged first, I'll update #34619.
- If #34619 is merged first, I'll update this.
- If #34619 is closed without merging, I'll leave this as is.